python/ray/tune/tests/execution/test_actor_caching.py

```diff
@@ -21,5 +23,108 @@ def test_actor_cached(tmpdir):
+    still an actor available for it to use (because it hasn't been evicted from
+    the cache, yet).
+
+    This test constructs such a situation an asserts that actor re-use does not
+    lead to errors in those cases.
+    """
+    tune_controller, actor_manger, resource_manager = create_execution_test_objects(
+        tmpdir, max_pending_trials=1
+    )
+    tune_controller._reuse_actors = True
+
+    assert not actor_manger.added_actors
+
+    trialA1 = TestingTrial(
+        "trainable1",
+        stub=True,
+        trial_id="trialA1",
+        experiment_path=str(tmpdir),
+        placement_group_factory=PlacementGroupFactory([{"CPU": 1}]),
+    )
+    tune_controller.add_trial(trialA1)
+    trialB1 = TestingTrial(
+        "trainable1",
+        stub=True,
+        trial_id="trialB1",
+        experiment_path=str(tmpdir),
+        placement_group_factory=PlacementGroupFactory([{"CPU": 5}]),
+    )
+    tune_controller.add_trial(trialB1)
+    trialA2 = TestingTrial(
+        "trainable1",
+        stub=True,
+        trial_id="trialA2",
+        experiment_path=str(tmpdir),
+        placement_group_factory=PlacementGroupFactory([{"CPU": 1}]),
+    )
+    tune_controller.add_trial(trialA2)
+    tune_controller.step()
+
+    # Prevent trial A3 from being staged by setting the number
+    # of pending actors to the maximum allowed
+    actor_manger.set_num_pending(2)
+
+    trialA3 = TestingTrial(
+        "trainable1",
+        stub=True,
+        trial_id="trialA3",
+        experiment_path=str(tmpdir),
+        placement_group_factory=PlacementGroupFactory([{"CPU": 1}]),
+    )
+    tune_controller.add_trial(trialA3)
+    tune_controller.step()
+
+    tracked_actorA1, _, _ = actor_manger.added_actors[0]
+    tracked_actorB1, _, _ = actor_manger.added_actors[1]
+    tracked_actorA2, _, _ = actor_manger.added_actors[2]
+
+    # Start trial A1, report that it's done training.
+    # This will cache the actor for A1 as A2 is already scheduled.
+    tune_controller._actor_started(tracked_actorA1)
+    tune_controller._on_training_result(trialA1, {"done": True})
+
+    # Trial A2 should be in the staged trials. A3 should still not be staged.
+    assert trialA2 in tune_controller._staged_trials
+    assert trialA3 not in tune_controller._staged_trials
+
+    # The actor of A1 should be cached for re-use now.
+    assert tune_controller._actor_cache.num_cached_objects == 1
+
+    # In the meantime, actor A2 started. This will unstage it.
+    tune_controller._actor_started(tracked_actorA2)
+
+    # Now, an external source (e.g. the BOHB scheduler) wants to prematurely
+    # stop trial A2. This will leave the cached actor intact, but trial A3
+    # is still not scheduled.
+    tune_controller._schedule_trial_stop(trialA2)
+    assert tune_controller._actor_cache.num_cached_objects == 1
+
+    # Process events. This will invoke "path 3" in TuneController._maybe_add_actors
+    # and re-use the cached actor
+    tune_controller.step()
```

Comment:
A3 is pending here, why doesn't path 2 get invoked in maybe_add_actors? Then, when `schedule_trial_actor` gets called, it also tries to reuse an actor.

I feel like it's easier to understand if all pending trials get started in that 2nd path, and then remove the 3rd path.

Reply:
In this case, we have max pending == 1, but then manually set num pending = 2 to trigger the 3rd path. When does this scenario happen normally -- is it only because BOHB will set trial statuses to pending without respecting the max pending trials?

I think that's why it gets skipped here:

```
                if self._actor_manager.num_pending_actors >= self._max_pending_trials:
                    break
```

Reply:
Decision: Let's try to simplify this after branch cut. Let's go with this fix that only includes minor changes for now.